omisego-sdk/src/main/java/co/omisego/omisego/websocket/listener/SocketCustomEventListener.kt

```diff
@@ -0,0 +1,120 @@
+
+import co.omisego.omisego.model.APIError
+import co.omisego.omisego.model.socket.SocketReceive
+import co.omisego.omisego.model.transaction.consumption.TransactionConsumption
+import co.omisego.omisego.websocket.event.SocketEvent
+import co.omisego.omisego.websocket.event.TransactionConsumptionFinalizedEvent
+import co.omisego.omisego.websocket.event.TransactionConsumptionRequestEvent
+import co.omisego.omisego.websocket.strategy.FilterStrategy
+
+interface SocketCustomEventListener {
+    var strategy: FilterStrategy
+    fun onEvent(event: SocketEvent<*>)
+
+    companion object {
+        /**
+         * A convenient method for listening for the specific event.
+         *
+         * @param lambda A lambda which receives the `SocketEvent` object without regarding the `SocketTopic`.
+         */
+        inline fun <reified Event : SocketEvent<*>> forEvent(
+            crossinline lambda: (Event) -> Unit
+        ): SocketCustomEventListener {
+            return object : SimpleSocketCustomEventListener<Event>() {
+                override var strategy: FilterStrategy = FilterStrategy.Event(listOf(Event::class.java))
+                override fun onSpecificEvent(event: Event) {
+                    lambda(event)
+                }
+            }
+        }
+
+        /**
+         * A convenient method for listening for any events, but the events will be filtered out by the provided `FilterStrategy`
+         *
+         * @param listener A [SocketCustomEventListener] implementation
+         * @param strategy A [FilterStrategy] that used for filtering an event.
+         */
+        fun forEvents(
+            listener: SocketCustomEventListener,
+            strategy: FilterStrategy
+        ): SocketCustomEventListener {
+            return object : SimpleSocketCustomEventListener<SocketEvent<*>>() {
+                override var strategy: FilterStrategy = strategy
+                override fun onSpecificEvent(event: SocketEvent<*>) {
+                    listener.onEvent(event)
+                }
+            }
+        }
+
+        /**
+         * A convenient method for listening for any events, but the events will be filtered out by the provided `FilterStrategy`.
+         *
+         * @param strategy A [FilterStrategy] that used for filtering an event.
+         * @param lambda A lambda which receives the `SocketEvent` object.
+         */
+        inline fun forEvents(
+            strategy: FilterStrategy,
+            crossinline lambda: (SocketEvent<out SocketReceive.SocketData>) -> Unit
+        ): SocketCustomEventListener {
+            return object : SimpleSocketCustomEventListener<SocketEvent<*>>() {
+                override var strategy: FilterStrategy = strategy
+                override fun onSpecificEvent(event: SocketEvent<*>) {
+                    lambda(event)
+                }
+            }
+        }
+    }
+
+    abstract class TransactionRequestListener : SimpleSocketCustomEventListener<SocketEvent<*>>() {
```

Comment:
I would move those out of this file, the file will bloat when the event number grows

Reply:
Yeah, it can be moved now because I changed to use `*TopicListener` for `Listenable.kt`. The old way will be deprecated as a compile error anyway.